ports/stm32f4/peripherals/stm32f4/pins.h

```diff
@@ -38,28 +38,32 @@
 
 typedef struct {
     mp_obj_base_t base;
-    //uint8_t number; //(3)port,(5)pin 
-    uint8_t port_number;
-    GPIO_TypeDef * port;
     uint8_t number;
+    uint8_t adc_num_input; //(3)mask ()
 } mcu_pin_obj_t;
 
-// extern GPIO_TypeDef *port_lookup_table[];
+#define ADC_1 	1
+#define ADC_123	7
+#define ADC_12 	3
+#define ADC_3 	4
 
-// static inline GPIO_TypeDef * get_GPIO_ptr(const mcu_pin_obj_t *p)
-// {
-// 	return port_lookup_table[0x7&( (p->number) >> 5)];
-// }
+//STM32 ADC pins can have a combination of 1, 2 or all 3 ADCs on a single pin,
+//but all 3 ADCs will share the same input number per pin. 
+#define ADC_INPUT(mask, number) \
+	.adc_num_input = (((mask) << 5) | ((number) & 0x1F)),
+
+#define NO_ADC \
+	.adc_num_input = 0xff,
 
 extern const mp_obj_type_t mcu_pin_type;
 
-// Used in device-specific pins.c
-#define PIN(p_port_num, p_port, p_number)       \
+// STM32 can have up to 9 ports, each restricted to 16 pins
+// We split the pin/port evenly, in contrast to nrf. 
+#define PIN(p_port, p_number, p_adc)       \
 { \
     { &mcu_pin_type }, \
-    .port_number = (p_port_num), \
-    .port = (p_port),      \
```

Comment:
I'd leave these separate and just mark the number of bits for each. It'll be clearer and will remove the need for pin_port and pin_mask everywhere. Instead, you'll be able to read the struct field directly.

Reply:
You're referring to the bit packing in this case, correct? I based this off the format of Dan's code in the nrf port - should we standardize this within nrf as well?

Reply:
Also, I won't necessarily reduce the references to pin_port and pin_mask, since I still need them for fetching the port address and performing the mask shift. I think acquiring the port address is unavoidable but I could simply perform the mask shift in every reference. Do you think that'd be preferable stylistically?

Reply:
Yup, referring to the bitpacking. I wouldn't port back changes to nrf now because we may evolve more here.

You are right about pin_port and pin_mask. It's preferable to still have them.

Reply:
Ok, will revert and fix